src/checkpoint/container/vector_serialize.h

```diff
@@ -53,29 +56,57 @@ namespace checkpoint {
 
 template <typename SerializerT, typename T, typename VectorAllocator>
 typename std::enable_if_t<
-  not std::is_same<SerializerT, checkpoint::Footprinter>::value,
-  void
-> serializeVectorMeta(SerializerT& s, std::vector<T, VectorAllocator>& vec) {
+  not std::is_same<SerializerT, checkpoint::Footprinter>::value, SerialSizeType
+>
+serializeVectorMeta(SerializerT& s, std::vector<T, VectorAllocator>& vec) {
   SerialSizeType vec_capacity = vec.capacity();
   s | vec_capacity;
   vec.reserve(vec_capacity);
 
   SerialSizeType vec_size = vec.size();
   s | vec_size;
-  vec.resize(vec_size);
+  return vec_size;
 }
 
 template <typename SerializerT, typename T, typename VectorAllocator>
 typename std::enable_if_t<
-  std::is_same<SerializerT, checkpoint::Footprinter>::value,
-  void
-> serializeVectorMeta(SerializerT& s, std::vector<T, VectorAllocator>& vec) {
+  std::is_same<SerializerT, checkpoint::Footprinter>::value, SerialSizeType
+>
+serializeVectorMeta(SerializerT& s, std::vector<T, VectorAllocator>& vec) {
   s.countBytes(vec);
+  return vec.size();
+}
+
+template <typename T, typename VectorAllocator>
+void constructVectorData(
+  SerialSizeType const vec_size, std::vector<T, VectorAllocator>& vec,
+  typename ReconstructorTraits<T>::template isDefaultConsType<T>* = nullptr
+) {
+  vec.resize(vec_size);
+}
+
+template <typename T, typename VectorAllocator>
+void constructVectorData(
+  SerialSizeType const vec_size, std::vector<T, VectorAllocator>& vec,
+  typename ReconstructorTraits<T>::template isNotDefaultConsType<T>* = nullptr
+) {
+  using Alloc = dispatch::Allocator<T>;
+  using Reconstructor =
+    dispatch::Reconstructor<typename dispatch::CleanType<T>::CleanT>;
+
+  Alloc allocated;
+  auto* reconstructed = Reconstructor::construct(allocated.buf);
+  vec.resize(vec_size, *reconstructed);
 }
```

Comment:
If `T` is not default-constructible reconstruct `T` using `Reconstructor` and resize vector.

Note - it uses copy ctor here. Would it make sense to have another function for `T` that's not copy-constructible but is move-constructible, and reconstruct vector using for-loop and `emplace_back()` + `std::move()`?

Reply:
Added.